RULES/xilinx/synth_out_of_context.go

```diff
@@ -0,0 +1,111 @@
+package xilinx
+
+import (
+	"dbt-rules/RULES/core"
+	"dbt-rules/RULES/hdl"
+	h "dbt-rules/hdl"
+	"fmt"
+)
+
+type ConstraintsFileScriptParams struct {
+	Out         core.Path
+	ClockSignal string
+	ClockPeriod float32
+}
+
+// Target for out-of-context synthesis run
+type SynthOutOfContext struct {
+	// Name of the top-level module to implement
+	Name string
+
+	// Top-level IP
+	Ip hdl.Ip
+
+	// Name of the clock input signal
+	ClockSignal string
+
+	// Target clock period (ns)
+	ClockPeriod float32
+
+	// Constraint definitions file for the design.
+	Constraints core.Path
+
+	// List of directories with board definitions
+	BoardFiles []core.Path
+}
+
+func (rule SynthOutOfContext) Build(ctx core.Context) {
+	ips := []core.Path{}
+	rtls := []core.Path{}
+	constrs := []core.Path{}
+
+	ins := []core.Path{}
+	for _, ip := range hdl.FlattenIpGraph([]hdl.Ip{rule.Ip}) {
+		for _, src := range ip.Sources() {
+			if hdl.IsRtl(src.String()) {
+				rtls = append(rtls, src)
+			} else if hdl.IsConstraint(src.String()) {
+				constrs = append(constrs, src)
+			} else if hdl.IsXilinxIpCheckpoint(src.String()) {
+				ips = append(ips, src)
+			}
+			ins = append(ins, src)
+		}
+	}
+
+	clockSignal := "clk_i"
+	clockPeriod := float32(1.550)
+
+	if rule.ClockSignal != "" {
+		clockSignal = rule.ClockSignal
+	}
+
+	if rule.ClockPeriod != 0.0 {
+		clockPeriod = rule.ClockPeriod
+	}
+
+	if rule.Constraints != nil {
+		ins = append(ins, rule.Constraints)
+		constrs = append(constrs, rule.Constraints)
+	} else {
+		// Use default out-of-context constraints - Meant for module analysis, NOT hierarchical design.
+		outConstr := ctx.Cwd().WithSuffix("/constraints.xdc")
```

Comment:
I would advise to incorporate the name of the top-level module into the name. That way you can have multiple of those in the same directory.